src/zampy/datasets/cds_utils.py

```diff
@@ -0,0 +1,371 @@
+from zampy.datasets import converter
+from zampy.datasets.dataset_protocol import SpatialBounds
+from zampy.datasets.dataset_protocol import TimeBounds
+from zampy.datasets.utils import get_file_size
+from zampy.reference.variables import VARIABLE_REFERENCE_LOOKUP
+
+
+PRODUCT_FNAME = {
+    "reanalysis-era5-single-levels": "era5",
+    "reanalysis-era5-land": "era5-land",
+    "cams-global-ghg-reanalysis-egg4": "cams",
+}
+SERVER_API = {
+    "era5": "cdsapi",
+    "era5-land": "cdsapi",
+    "cams": "adsapi",
+}
+CONFIG_PATH = Path.home() / ".config" / "zampy" / "zampy_config.yml"
+CDSAPI_FALLBACK_PATH = Path.home() / ".cdsapirc"
+
+
+def cds_request(
+    dataset: str,
+    variables: list[str],
+    time_bounds: TimeBounds,
+    spatial_bounds: SpatialBounds,
+    path: Path,
+    cds_var_names: dict[str, str],
+    overwrite: bool,
+) -> None:
+    """Download data via CDS API.
+
+    To raise a request via CDS API, the user needs to set up the
+    configuration file `.cdsapirc` following the instructions on
+    https://cds.climate.copernicus.eu/api-how-to.
+
+    If the target dataset is available on ADS, then please set up
+    the configuration file `.cdsapirc` following the instructions on
+    https://ads.atmosphere.copernicus.eu/api-how-to.
+
+    It is recommended to put your ADS API key and CDS API key to the zampy
+    config file at `CONFIG_PATH`.
+
+    Following the efficiency tips of request for ERA5 and ERA5-land dataset,
+    https://confluence.ecmwf.int/display/CKB/Climate+Data+Store+%28CDS%29+documentation
+    The downloading is organized by asking for one month of data per request.
+
+    Args:
+        dataset: Dataset name for retrieval via `cdsapi`.
+        variables: Zampy variables.
+        time_bounds: Zampy time bounds object.
+        spatial_bounds: Zampy spatial bounds object.
+        path: File path to which the data should be saved.
+        cds_var_names: Variable names from CDS server side.
+        overwrite: If an existing file (of the same size!) should be overwritten.
+    """
+    fname = PRODUCT_FNAME[dataset]
+
+    url, api_key = cds_api_key(fname)
+
+    c = cdsapi.Client(
+        url=url,
+        key=api_key,
+        verify=True,
+        quiet=True,
+    )
+    # choose retrieve function
+    retrieve_func = RETRIEVE_FUNCTION[fname]
+
+    # start datasets retrieval
+    retrieve_func(
+        c,
+        fname,
+        dataset,
+        variables,
+        time_bounds,
+        spatial_bounds,
+        path,
+        cds_var_names,
+        overwrite,
+    )
+
+
+def cds_api_key(product_name: str) -> tuple[str, str]:
+    """Load url and CDS/ADS API key.
+
+    This function loads url and CDS/ADS API key from zampy config file.
+
+    If zampy config file is not available, this function will look for
+    `.cdsapirc` as fallback.
+
+    Args:
+        product_name: Dataset name alias.
+
+    Returns:
+        url and API key for making a request to CDS server.
+    """
```

Comment:
In this PR, we load CDS/ADS API key from zampy config. When it is not available, this function will look for the fallback solution, which is the `.cdsapirc` at user's home path. The implementation is not perfect for now, given that there are many checks and exceptions that need to be considered (e.g. whether key is empty, url is valid , etc.).

I think later we need a more dedicated function for this. But for now I would suggest to keep it like this, otherwise this PR just gets toooo big 😹.

Reply:
I don't think we need to do too many checks, the key will just be declined by the cdsapi with a nice error.

This fallback solution might get a bit tricky, as the URL is just used to either point to ADS or CDS, while we know to which one we want to connect.

My preference is to have no fallback, except perhaps checking to see if the file is there and tell the user "Found .cdsapirc file. Please use these keys to setup the zampy configuration. For more information see ...".

Reply:
Sounds good to me. This also makes the implementation easier. Then let's ""push" the user to live with the best zampy practice 😎 .